chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2479 @@
+            raise Exception("Error reading block %04x." % (start))
+
+        block_data = response[4:-1]
+        # print("block_data = {}".format(str(binascii.b2a_hex(block_data))))
+
+    except:
+        raise errors.RadioError("Failed to read block at %04x" % start)
+
+    return block_data
+
+
+def _get_radio_firmware_version(radio):
+    if radio.MODEL in ALL_MODEL:
+        block = _read_block(radio, 0x1B40, 0x20)
+        version = block[0:6]
+    return version
+
+
+IDENT_BLACKLIST = {
+    b"\x50\x56\x4F\x4A\x48\x1C\x14": "Radio identifies as TIDRADIO TD-H8",
+}
+
+
+def _ident_radio(radio):
+    for magic in radio._idents:
+        error = None
+        try:
+            data = _do_ident(radio, magic)
+            return data
+        except errors.RadioError as e:
+            # LOG.error("tdh8._ident_radio: %s", e)
+            error = e
+            time.sleep(2)
+
+    if error:
+        raise error
+    raise errors.RadioError("Radio did not respond")
+
+
+def _do_download(radio):
+    data = _ident_radio(radio)
+    # radio_version = _get_radio_firmware_version(radio)
+    # LOG.info("Radio Version is %s" % repr(radio_version))
+    append_model = False
+    # HAM OR GMRS
+    LOG.info("Radio mode is " + MODE_RADIO)
+    LOG.info("Chirp choose mode is " + radio.ident_Mode)
+    # The Ham and Gmrs modes are subclasses of this model tdh8.
+    # When the Ham or Gmrs mode is selected,
+    # a double verification operation will be performed to
+    # obtain the current mode code of the walkie-talkie and
+    # the mode code currently selected by the chirp for comparison.
+    # If the two are consistent, enter This mode,
+    # otherwise an inconsistent error will be prompted.
+    if MODE_RADIO == radio.ident_Mode:
+        LOG.info("Successful match.")
+    else:
+        msg = (
+            "Please check whether the walkie-talkie mode is consistent with "
+            "the selected model mode.")
+        raise InvalidValueError(msg)
+
+    # Main block
+    LOG.info("Downloading...")
+
+    for i in range(0, radio._memsize, 0x20):
+        block = _read_block(radio, i, 0x20)
+        data += block
+        _do_status(radio, i)
+    _do_status(radio, radio._memsize)
+    LOG.info("done.")
+
+    if append_model:
+        data += radio.MODEL.ljust(8)
+
+    # LOG.debug("done.")
+    return memmap.MemoryMapBytes(data)
+
+
+def _exit_write_block(radio):
+    serial = radio.pipe
+    try:
+        serial.write(b"E")
+
+    except:
+        raise errors.RadioError("Radio refused to exit programming mode")
+
+
+def _write_block(radio, addr, data):
+    serial = radio.pipe
+    cmd = struct.pack(">cHb", b'W', addr, 0x20)
+    data = radio.get_mmap()[addr + 8: addr + 40]
+    cmd_str = hex(sum(data) & 0xff)
+    data_bused = chr(int(cmd_str, 16))
+
+    # This is because when our cmd_str is converted to decimal,
+    # sometimes num will be greater than 128, but chr(num).encode('utf-8') gets
```

Comment:
This is why you should not be encoding strings to `bytes` at all. You want something from `struct.pack()` if you're building something up, or use `bytes()` itself.  What you have on L742 and L743 is not the proper way to do this, as `int(hex(value), 16)` is an identity operation - it just wastes cpu. This is all you need for the checksum stuff above:
```
data_bused = bytes([sum(data) & 0xFF])
```

But also, "data_bused" does not mean anything to me. I think you want "checksum" ?

Reply:
Yes, a checksum is required here, the driver needs a checksum to verify. I will add a comment